### PR TITLE
Add model for interaction between Lucene & Version Map on replicas

### PR DESCRIPTION
**Pull request by DaveCTurner**
This adds a model for how we should track sequence numbers for indexing and deletions in the replica, using the Version Map and Lucene. It includes an optimisation that can avoid the need to perform these (relatively expensive) checks for append-only operations when it is safe to do so.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Summary of verbal review with @bleskes:

The document versions are not relevant for the replica - they are a relic of older models that modelled more of the primary too. Removed in 0165d83.

However, the model of Lucene needs refinement: we model replacement of a document as if calling the update API, but there is also an add API which doesn't delete any previous versions first. The add API can only be used if we're sure there is nothing that needs removing, but is more efficient so should be used where possible. The add API can be modelled as having the same effect as the update API but with an additional assertion that there is no document to overwrite.

**Comment by DaveCTurner** [comment]
@dnhatn it's likely you will be involved in the implementation of this so it might be useful for you to contemplate this PR too. Not sure what level you're at with this kind of modelling but I have time to walk you through it if needed.

**Comment by dnhatn** [comment]
@DaveCTurner Thanks for pinging me. I think I will focus on the implementation. I will surely have a look on this but mainly for learning purpose.

File: replica-version-map/tla/versionmap.tla [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+
+            if replication_request.seqno > local_check_point + 1
+            then
+                (* cannot advance local check point *)
+                seqnos_above_local_check_point := seqnos_above_local_check_point
+                    \union {replication_request.seqno};
+
+            else
+
+                AdvanceLocalCheckPoint:
```
Comment (DaveCTurner) [inline replica-version-map/tla/versionmap.tla]: Do we want the advancing of the local checkpoint to be a separate process?

File: replica-version-map/tla/versionmap.tla [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+Next == ReplicationRequestGeneratorProcess \/ LuceneProcess
+           \/ ReplicaEngineProcess
+           \/ (* Disjunct to prevent deadlock on termination *)
+              ((\A self \in ProcSet: pc[self] = "Done") /\ UNCHANGED vars)
+
+Spec == Init /\ [][Next]_vars
+
+Termination == <>(\A self \in ProcSet: pc[self] = "Done")
+
+\* END TRANSLATION
```
Comment (DaveCTurner) [inline replica-version-map/tla/versionmap.tla]: Do we want some more invariants below? At the moment they're mostly about properties at termination, which is simple and ought to be sufficient, but we could have more if desired.

**Comment by DaveCTurner** [comment]
Closing this in favour of #29.

**Pull request closed without merging**